### PR TITLE
EIP 2619 - Geotimeline Contact Tracing Data Standard

### PR DESCRIPTION
**Pull request by marleymarl**
---
eip: 2619
title: Geotimeline Contact Tracing Data Standard
author: Greg Marlin (@marleymarl)
discussions-to: [https://github.com/marleymarl/geotime/issues/95](https://github.com/marleymarl/geotime/issues/95)
status: Draft
type: Standards Track
category: ERC
created: 2020-04-23
---

## Simple Summary

A standard interface for storing to and retrieving from the blockchain historical points of location and time, collectively those points forming a geotimeline for individuals, for the purpose of contact tracing for COVID-19 and any other pandemic, in an anonymous and private way, while at the same time ensuring data integrity. 

## Abstract

This standard outlines a smart contract interface for recording and retrieving geotimelines for the purpose of contact tracing. The key issue is that it is important to share this data and at the same time maintain privacy and anonymity of both confirmed cases of COVID and of those people wanting to check their geotimelines against that data. At the same time it is important to provide a way to ensure that the data is entered through trusted parties. The Geotimeline standard enables contact tracing organizaitons to be whitelisted before securely, anonymously and privately storing this information and making it available for retrieval. 

## Motivation

Contact tracing is a critical activity for reducing the spread of COVID-19 and other pandemic viruses, and in helping to mitigate risks after the end of quarantine periods. Key issues related to contact tracing include data integrity and availability, as well as individual privacy and anonymity. 

The [GeoTimeline](https://geotimeline.com) international team of volunteers have been working on an open source map data input front-end [component library](https://github.com/marleymarl/geotime) that makes it easy to capture geotimelines, points of location and time that each represent a set of historical footprints that identify to one person but protect privacy and anonymity through the assignment of random uuids. 

Contact tracing workers capture this information by clicking on a map and adding a date and time to associate with that latitude and longitude point using a pop-up modal form. After a set of these points with date and time are collected, they are saved as one geotimeline.

The [GeoTimeline Application](https://app.geotimeline.com) can be used by contact tracing organization that are typically made up of many human team members interviewing people and recording the information.

This standard addresses the issues of privacy and anonymity as well as data integrity and availability in the following ways: 

* Contact tracing organizations are added to a whitelist to add geotimeline data to the blockchain using their public wallet addresses.
* Thus only whitelisted organization addresses can sign transactions to add data
* Privacy and anonymity of the individual people whose whereabouts and status are represented by the geotimeline are protected through no Personally Identifiable Information being associated with the geotimeline. Each geotimeline is made up only of the data points of date/time and lat/long points. The identifier is the id generated for the geotimeline on the blockchain, which is the index of the geotimeline in the geotimelines array. 

By applying this standard, GeoTimeline will integrate secure signing and authentication of a trusted party through the integration of an Ethereum wallet on the frontend (using MetaMask / web3.js) to be used to prove the organization entering the data is a verified organization. The geotimelines will be represented on the blockchain as an address with associated geotimeline bounding box coordinates and status (confirmed case, suspect, not suspect, recovered). Nothing will be able to be tied back to an individual's personal identity by looking at the blockchain. At the same time they key information that is needed to be open and available will be there and queriable, namely the geotimeline information. 

#### geotimeline

Returns the hash of set of footprints associated with a timeline address, along with it geotimeline bounding box values (2 geohashes representing top-left and bottom-right lat/lon values and 2 timestamps representing earliest time and latest time) and case status

``` js
function geotimeline(uint geotimelineId) public view returns(string geotimelineHash, uint earliestTime, uint latestTime, string upperLeftGeoHash, string bottomRightGeoHash, string status)
```

#### getAllGeoTimelines

Returns all timeine addresses that fall within geotimeline bounding box values (timestamps falling in between earliestTime and lastestTime and geohashes falling in between upperLeftGeoHash and bottomRightGeoHash)

``` js
function getAllGeoTimelines(uint earliestTime, uint latestTime, string upperLeftGeoHash, string bottomRightGeoHash) public view returns(GeoTimeline[], geotimelines)
```

#### addGeoTimeline

Saves a timeline using an IPFS hash representing all footprints in a geotimeline, along with its bounding box values of start date/time, end date/time, upper-left geohash and bottom-right geohash as well as status (confired case, not suspect, suspect, recovered) and gets back an Id for the geotimeline that was saved to the array

``` js
function addGeoTimeline(string geotimelineHash, uint earliesttime, uint latesttime, string upperLeftGeoHash, string bottomRightGeoHash, string status) public onlyVerifiedOrganization(msg.sender) returns(uint geotimelineId)
```

#### addVerifiedOrganization

Adds the wallet address of a contact tracing organization to a list of approved addresses that can add geotimeline data

``` js
function addVerifiedOrganization(address newVerifiedOrganization, string _name) public onlyOwner returns(uint verifiedOrganizationId)
```

#### isVerifiedOrganization

Adds the wallet address of a contact tracing organization to a list of approved addresses that can add geotimeline data

``` js
function isVerifiedOrganization(address verifiedOrganization) public constant returns(bool)
```

#### updateTimelineStatus

Updates the status of a given timeline to reflect a new value (non suspect, suspect, confirmed, recovered)

``` js
function updateTimelineStatus(uint geotimelineId, string newStatus) public returns(bool success)
```

#### GeoTimelineAdded

MUST trigger when a new geotimeline is added. 

``` js
event GeoTimelineAdded(uint geotimelineId)
```

#### VerifiedOrganizationAdded
 
MUST trigger when a new verifiedOrganization is added. 

``` js
event VerifiedOrganizationAdded(address verifiedOrganization)
```

#### VerifiedOrganizationRemoved

MUST trigger when a verifiedOrganization is removed. 

``` js
event VerifiedOrganizationRemoved(address verifiedOrganization)
```

#### ConfirmedCaseUpdate

MUST trigger when the status of a geotimeine is updated to confirmed.

``` js
event ConfirmedCaseUpdate(uint geotimelineId, string status);
```

## Rationale

The proposed solution was designed to provide a way to address key issues of data integrity and availability, while at the same time ensuring anonymity and privacy of the individuals whose geotimelines are being stored on and retrieved from the blockchain. 

By creating an EIP, we hope to increase the use of Ethereum as an open immutable data store for contact tracing data and encourage contact tracing organizations to thus make this data open and available while preserving privacy and anonymity. 

## Test Cases

GeoTimeline will be implementing this within its project and will be approaching contact tracing organizations to use it within the framework of the overall contact tracing solution. This section will be updated when GeoTimeline has deployed a smart contract based on this standard to the Ropsten test network. 

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

### CODE REVIEW

File: EIPS/eip-2611.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+---
+eip: 2611
```
Comment (axic) [inline EIPS/eip-2611.md:2]: Where is 2611 coming from? That seems to be something different: https://github.com/ethereum/EIPs/pull/2611

Comment (marleymarl) [reply EIPS/eip-2611.md]: Got it - thank you. Updated.

File: EIPS/eip-2619.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+GeoTimeline will be implementing this within its project and will be approaching contact tracing organizations to use it within the framework of the overall contact tracing solution. This section will be updated when GeoTimeline has deployed a smart contract based on this standard to the Ropsten test network. 
+
+
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+## Copyright
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:151]:
```suggestion
## Security Considerations
TBD

## Copyright
```

File: EIPS/eip-2619.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+By creating an EIP, we hope to increase the use of Ethereum as an open immutable data store for contact tracing data and encourage contact tracing organizations to thus make this data open and available while preserving privacy and anonymity. 
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+GeoTimeline will be implementing this within its project and will be approaching contact tracing organizations to use it within the framework of the overall contact tracing solution. This section will be updated when GeoTimeline has deployed a smart contract based on this standard to the Ropsten test network. 
+
+
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:149]:
```suggestion
TBD
```

File: EIPS/eip-2619.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+```
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+The proposed solution was designed to provide a way to address key issues of data integrity and availability, while at the same time ensuring anonymity and privacy of the individuals whose geotimelines are being stored on and retrieved from the blockchain. 
+
+By creating an EIP, we hope to increase the use of Ethereum as an open immutable data store for contact tracing data and encourage contact tracing organizations to thus make this data open and available while preserving privacy and anonymity. 
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:144]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+#### ConfirmedCaseUpdate
+
+MUST trigger when the status of a geotimeine is updated to confirmed.
+
+``` js
+event ConfirmedCaseUpdate(uint geotimelineId, string status);
+```
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:138]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+By applying this standard, GeoTimeline will integrate secure signing and authentication of a trusted party through the integration of an Ethereum wallet on the frontend (using MetaMask / web3.js) to be used to prove the organization entering the data is a verified organization. The geotimelines will be represented on the blockchain as an address with associated geotimeline bounding box coordinates and status (confirmed case, suspect, not suspect, recovered). Nothing will be able to be tied back to an individual's personal identity by looking at the blockchain. At the same time they key information that is needed to be open and available will be there and queryable, namely the geotimeline information. 
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+
+
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:44]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+This standard outlines a smart contract interface for recording and retrieving geotimelines for the purpose of contact tracing. The key issue is that it is important to share this data and at the same time maintain privacy and anonymity of both confirmed cases of COVID and of those people wanting to check their geotimelines against that data. At the same time it is important to provide a way to ensure that the data is entered through trusted parties. The Geotimeline standard enables contact tracing organizaitons to be whitelisted before securely, anonymously and privately storing this information and making it available for retrieval. 
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:23]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+A standard interface for storing to and retrieving from the blockchain historical points of location and time, collectively those points forming a geotimeline for individuals, for the purpose of contact tracing for COVID-19 and any other pandemic, in an anonymous and private way, while at the same time ensuring data integrity. 
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:19]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+author: Greg Marlin (@marleymarl)
+discussions-to: https://github.com/marleymarl/geotime/issues/95
+status: Draft
+type: Standards Track
+category: ERC
+created: 2020-04-23
+---
+
+<!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:13]:
```suggestion
```

File: EIPS/eip-2619.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
```
Comment (MicahZoltu) [inline EIPS/eip-2619.md:15]:
```suggestion
```

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
It looks like for some reason this change touches some files in the root of this repository.  Can you remove those changes so this can potentially be merged?

It is unclear to me why this needs to be a standard?  Keep in mind that not all good ideas should be standards, and while this may be a good idea, I think it would make more sense to just build an application/contract rather than trying to define a standard.

File: EIPS/eip-2619.md (hunk 2 above)
Comment (marleymarl) [reply EIPS/eip-2619.md:151]: Will look into this today.

**Pull request closed without merging**